modules/ppcp-api-client/src/Exception/PayPalApiException.php

```diff
@@ -141,4 +144,39 @@ public function get_details( string $error ): string {
 
 		return $details;
 	}
+
+	/**
+	 * Adds an improved error message to the response if the error detail is known.
+	 *
+	 * @param stdClass $json The response.
+	 * @return stdClass
+	 */
+	public function add_improved_error_message( stdClass $json ): stdClass {
+		if ( ! isset( $json->details ) ) {
+			return $json;
+		}
+		$improved_keys_messages = array(
+			'PAYMENT_DENIED'           => __( 'PayPal rejected the payment. Please reach out to the PayPal support for more information.', 'woocommerce-paypal-payments' ),
+			'TRANSACTION_REFUSED'      => __( 'The transaction has been refused by the payment processor. Please reach out to the PayPal support for more information.', 'woocommerce-paypal-payments' ),
+			'DUPLICATE_INVOICE_ID'     => __( 'The transaction has been refused because the Invoice ID already exists. Please create a new order or reach out to the store owner.', 'woocommerce-paypal-payments' ),
+			'PAYER_CANNOT_PAY'         => __( 'There was a problem processing this transaction. Please reach out to the store owner.', 'woocommerce-paypal-payments' ),
+			'PAYEE_ACCOUNT_RESTRICTED' => __( 'There was a problem processing this transaction. Please reach out to the store owner.', 'woocommerce-paypal-payments' ),
+		);
+		$improved_errors        = array_filter(
+			array_keys( $improved_keys_messages ),
+			function ( $key ) use ( $json ): bool {
+				foreach ( $json->details as $detail ) {
+					if ( isset( $detail->issue ) && $detail->issue === $key ) {
+						return true;
+					}
+				}
+				return false;
+			}
+		);
+		if ( $improved_errors ) {
+			$improved_errors = array_values( $improved_errors );
+			$json->message   = $improved_keys_messages[ $improved_errors[0] ];
```

Comment:
Maybe it would be better to just return the message instead of changing the response?

```
if ( $improved_errors ) {
    ...
    return $new_message;
}
return $response->message;
```

```
$message           = $this->get_customer_friendly_message($response);
```

Reply:
yeah, it makes more sense to change only the message, I was thinking we were sending back the object so I was trying to change it, but we send the response and the message. Now I see, thanks for the hint.